rust/noosphere-cli/tests/peer_to_peer.rs

```diff
@@ -83,118 +84,145 @@ async fn start_name_system_server(ipfs_url: &Url) -> Result<(JoinHandle<()>, Url
 
     Ok((
+        url,
         tokio::spawn(async move {
             let mut network = NameSystemNetwork::generate(2, store).await.unwrap();
             let node = network.nodes_mut().pop().unwrap();
             start_name_system_api_server(Arc::new(node), listener)
                 .await
                 .unwrap();
         }),
-        url,
     ))
 }
 
-#[tokio::test]
-async fn gateway_publishes_and_resolves_petnames_configured_by_the_client() {
-    initialize_tracing(None);
+struct SphereStack {
+    pub identity: Did,
+    pub workspace: Workspace,
+    _key_name: String,
+    _temp_dirs: (tempfile::TempDir, tempfile::TempDir),
+}
 
-    let ipfs_url = Url::parse("http://127.0.0.1:5001").unwrap();
-    let (ns_task, ns_url) = start_name_system_server(&ipfs_url).await.unwrap();
+struct FullStackSphere {
+    pub client: SphereStack,
+    pub gateway: SphereStack,
+    pub ns_url: Url,
+    pub ipfs_url: Url,
+    gateway_url: Option<Url>,
+    gateway_task: Option<JoinHandle<()>>,
+}
 
-    let (gateway_workspace, _gateway_temporary_directories) = Workspace::temporary().unwrap();
-    let (client_workspace, _client_temporary_directories) = Workspace::temporary().unwrap();
+impl FullStackSphere {
+    pub async fn new(name: &str, ipfs_url: &Url, ns_url: &Url) -> Result<Self> {
+        let (client_workspace, client_temp_dirs) = Workspace::temporary().unwrap();
+        let (gateway_workspace, gateway_temp_dirs) = Workspace::temporary().unwrap();
+        let client_key_name = format!("{}-CLIENT_KEY", name);
+        let gateway_key_name = format!("{}-GATEWAY_KEY", name);
+        key_create(&client_key_name, &client_workspace).await?;
+        key_create(&gateway_key_name, &gateway_workspace).await?;
+        sphere_create(&client_key_name, &client_workspace).await?;
+        sphere_create(&gateway_key_name, &gateway_workspace).await?;
+        let client_identity = client_workspace.sphere_identity().await.unwrap();
+        let gateway_identity = gateway_workspace.sphere_identity().await.unwrap();
+
+        config_set(
+            ConfigSetCommand::Counterpart {
+                did: client_identity.clone().into(),
+            },
+            &gateway_workspace,
+        )
+        .await?;
+        let client = SphereStack {
+            _key_name: client_key_name,
+            identity: client_identity,
+            workspace: client_workspace,
+            _temp_dirs: client_temp_dirs,
+        };
+        let gateway = SphereStack {
+            _key_name: gateway_key_name,
+            identity: gateway_identity,
+            workspace: gateway_workspace,
+            _temp_dirs: gateway_temp_dirs,
+        };
+        Ok(FullStackSphere {
+            client,
+            gateway,
+            gateway_url: None,
+            gateway_task: None,
+            ipfs_url: ipfs_url.to_owned(),
+            ns_url: ns_url.to_owned(),
+        })
+    }
+
+    pub async fn start_gateway(&mut self) -> Result<()> {
```

Comment:
Might be nice to document these interfaces just to make it easy for a newcomer to grok the test flow.